### PR TITLE
[battery] Linux support added in battery

### PR DESCRIPTION
**Pull request by yash1200**
## Description

In this PR, I have implemented the battery plugin for Linux. The information is taken from Dbus.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by yash1200** [comment]
Hey @ditman kindly have a look at this one.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (ditman) [inline packages/battery/battery_linux/lib/battery_linux.dart:26]: Would it be safer to enumerate the subdirectories of `/sys/class/power_supply` and grab one from there (or maybe control multiple BATx if more than one exists?)

**Review by ditman — COMMENTED** [review]
(Just a couple of general comments in the implementation of the plugin)

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (yash1200) [reply packages/battery/battery_linux/lib/battery_linux.dart:26]: It means we have to return a List of `batteryData` to display info of all the batteries.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (yash1200) [reply packages/battery/battery_linux/lib/battery_linux.dart:52]: I think `watcher` doesn't work with files in `root` directory. It's working with files in `/home` directory but not in `root` directory.

File: packages/battery/battery_linux/LICENSE
Comment (stuartmorgan) [inline packages/battery/battery_linux/LICENSE:3]: Please sync with master and re-copy this LICENSE file from one of the files in the repository; I just standardized the formatting of all of them (it's not yet enforced by any automated checks).

File: packages/battery/battery_linux/LICENSE
Comment (stuartmorgan) [inline packages/battery/battery_linux/LICENSE:1]: The Flutter Authors

This applies throughout the PR. (The current files are inconsistent, which I'm working on fixing, but there's no reason for new files to use legacy author lines.)

File: packages/battery/battery_linux/example/lib/main.dart
Comment (stuartmorgan) [inline packages/battery/battery_linux/example/lib/main.dart:8]: The battery_linux plugin should really use/test battery_linux directly; this is something I'm trying to move toward for all federated plugins (at least on desktop) going forward; duplicating main-plugin tests doesn't help since those will be run anyway, and there's no need to create circular dependencies. The platform interface implementation should provide a clear API surface to write examples and tests against.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (stuartmorgan) [reply packages/battery/battery_linux/lib/battery_linux.dart:52]: Is there a reason you aren't using D-Bus to get and monitor battery information? I'm not familiar with the details of the APIs, but a quick look suggest there are change notifications.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/battery/battery_linux/example/lib/main.dart
Comment (yash1200) [reply packages/battery/battery_linux/example/lib/main.dart:7]: Done.

File: packages/battery/battery_linux/LICENSE
Comment (yash1200) [reply packages/battery/battery_linux/LICENSE]: Done

File: packages/battery/battery_linux/LICENSE
Comment (yash1200) [reply packages/battery/battery_linux/LICENSE]: Done.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (yash1200) [reply packages/battery/battery_linux/lib/battery_linux.dart]: I haven't used D-Bus before. I tried to implement it and got to [this](https://gist.github.com/yash1200/9e0602eb2328b1ea975349b61b9aca4e). I couldn't figure out the part to implement change notifications in it. Can you guide me for it or provide source to implement it ?

Comment (stuartmorgan) [reply packages/battery/battery_linux/lib/battery_linux.dart]: I'm not going to be able to help with the details of using D-Bus; perhaps @robert-ancell can point you in the right direction.

Comment (yash1200) [reply packages/battery/battery_linux/lib/battery_linux.dart]: Isn't there any other method to watch over files? Just curious.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (yash1200) [reply packages/battery/battery_linux/lib/battery_linux.dart:26]: I think a better information is stated [here](https://askubuntu.com/a/309146/1050100). What's your call on this @stuartmorgan .

Comment (stuartmorgan) [reply packages/battery/battery_linux/lib/battery_linux.dart:26]: I'm not sure what your question is exactly, but I would not expect an answer to the question "How do I check the battery's status **via the terminal**?" to be a good source of information about how the best way to programmatically check, and monitor, battery information.

File: packages/battery/battery_linux/lib/battery_linux.dart
Comment (stuartmorgan) [reply packages/battery/battery_linux/lib/battery_linux.dart]: @yash1200 I don't understand why you want to do this by watching files in the first place rather than using an API, so I don't have a comment on that.

**Comment by yash1200** [comment]
Hey @stuartmorgan kindly have a look at these changes.

**Comment by ditman** [comment]
@yash1200 I think there's a better way for this contribution to be landed. I'm talking to people in the flutter community where they're working in a very similar effort (/cc @slightfoot + @jpnurmi). If you all agree, we can close this PR and get Yash onboarded in the community effort!

**Comment by yash1200** [comment]
@ditman Is there any way to contact you all and discuss regarding this as this has been on hold for a while now.

**Comment by stuartmorgan** [comment]
@ditman Is there a reason we wouldn't want a Linux implementation of this plugin in the repository with all the other platforms?

@robert-ancell Can you provide a pointer to the package that wraps the battery API?

**Comment by robert-ancell** [comment]
> wrap

**Comment by robert-ancell** [comment]
> @robert-ancell Can you provide a pointer to the package that wraps the battery API?

Do you mean https://pub.dev/packages/upower ?

**Comment by stuartmorgan** [comment]
> > @robert-ancell Can you provide a pointer to the package that wraps the battery API?
> 
> Do you mean https://pub.dev/packages/upower ?

Yes, thanks! We'll need to wait for offline license discussions to resolve though (for either this or `dbus`).

**Comment by ditman** [comment]
I'm going to close this one, as work for the Linux Battery plugin is underway elsewhere. I've already talked to @yash1200 about this!

Thanks for the contribution @yash1200, keep them coming!

**Pull request closed without merging**
